tfjs-backend-webgpu/src/webgpu_program.ts

```diff
@@ -50,8 +50,285 @@ export interface WebGPUProgram {
   getUserCode: () => string;
 }
 
-export interface TensorData {
-  dtype: DataType;
+export const compileProgram =
+    (device: GPUDevice, program: WebGPUProgram,
+     pipelineLayout: GPUPipelineLayout, inputsData: InputInfo[],
+     output: TensorInfo, isFromPixel = false): GPUComputePipeline => {
+      const outputData = {dtype: output.dtype, shape: output.shape};
+      const source = makeShader(inputsData, outputData, program, isFromPixel);
+      const module = device.createShaderModule(
+          {code: source, label: program.constructor.name});
+      const pipeline = device.createComputePipeline({
+        layout: pipelineLayout,
+        compute: {module, entryPoint: 'main'},
+        label: program.constructor.name
+      });
+
+      return pipeline;
+    };
+
+export function getCoordsDataType(rank: number): string {
+  if (rank <= 1) {
+    return 'i32';
+  } else if (rank === 2) {
+    return `vec2<i32>`;
+  } else if (rank === 3) {
+    return `vec3<i32>`;
+  } else if (rank === 4) {
+    return `vec4<i32>`;
+  } else if (rank === 5) {
+    return `vec5`;
+  } else if (rank === 6) {
+    return `vec6`;
+  } else {
+    throw Error(`GPU for rank ${rank} is not yet supported`);
+  }
+}
+
+export function getCoordsXYZ(index: number): string {
+  if (index === 0) {
+    return 'x';
+  } else if (index === 1) {
+    return 'y';
+  } else if (index === 2) {
+    return 'z';
+  } else if (index === 3) {
+    return 'w';
+  } else if (index === 4) {
+    return 'u';
+  } else if (index === 5) {
+    return 'v';
+  } else {
+    throw Error(`Index ${index} is not yet supported`);
+  }
+}
+
+export function getInputByOutputSnippet(
```

Comment:
It seems that `getInputByOutputSnippet` should be a private function. `export` is not needed. Please also check other places.

Reply:
done